### PR TITLE
Support bindings for literal defines

### LINKED ISSUE #4: Support bindings for defines
Certain symbols in the standard C library is exposed under the expected names using a `#define` rather than an `extern` global variable, e.g. `stdin` from `stdio.h`:

```c
extern FILE *__stdinp;
// ...
#define stdin   __stdinp
```

The way this is handled in Scala Native is to generate a C function which simply returns the expected symbol, e.g.:

```c
void *scalanative_libc_stdin() { return stdin; }
```

which is then bound using

```scala
@extern
object stdio {
  // ...
  @name("scalanative_libc_stdin")
  def stdin: Ptr[FILE] = extern
}
```

The questions is should the binding generator handle this case by generating C code or should it generate per platform bindings that use the final symbol, e.g.:

```scala
@extern
object stdio {
  // ...
  @name("__stdinp")
  def stdin: Ptr[FILE] = extern
}
```

### PR DESCRIPTION
**Pull request by kornilova-l**
Part of #4 

Currently the code only finds defines and does not generate any code for it.

I think it would be better to implement end-to-end testing before moving to this one #51

### CODE REVIEW

**Comment by kornilova-l** [comment]
I decided that literals will be places directly to Scala code, so their values will be known by a programmer without looking into header file (see [Support Bindings for Defines Design Doc](https://github.com/kornilova-l/scala-native-bindgen/wiki/Support-Bindings-for-Defines-Design-Doc))

This commit will be only about literals.

Currently there are following issue:
* We need to find a way to get Scala type for `numeric_constant` token.
    For example, we need to know that this token `10000000000` is `Long`, so we can correctly use it in a Scala file:
    ```scala
    object DefineDefines {
      val LONG: native.CLong = 10000000000L
    }
    ```
* Find a way to follow chain of defines.
    Example:
    ```c
    #define INT 42
    #define INT_2 INT
    ```
* Check that defines are not `undef` below in the header

**Comment by kornilova-l** [comment]
Last commit is incorrect, I'll fix it tomorrow

File: bindgen/defines/DefineFinder.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                llvm::errs().flush();
+            }
+        } else if (finalToken->isAnyIdentifier()) {
+            // TODO: save identifier and get its type in ScalaFrontendAction
+        }
+    }
+}
+
+const clang::Token *
+DefineFinder::getFinalIdentifier(const clang::Token &token) const {
```
[the remark this answers is not in the document]
Comment (kornilova-l) [reply bindgen/defines/DefineFinder.cpp]: Name is good
It should be called recursively until the identifier is not an another define, but the problem is that defines may contain multiple tokens:
```c
extern int a;
#define MY_A a
#define MY_A_PTR MY_A * // two tokens, one of them is another define
```
So the function might get too complicated. I'll try to find clang function that will resolve value of define for us

File: bindgen/ir/IR.h [hunk 2]
```diff
@@ -28,6 +29,11 @@ class IR {
     void addUnion(std::string name, std::vector<Field> fields,
                   uint64_t maxSize);
 
+    void addLiteralDefine(std::string name, std::string literal);
+
+    void addLiteralDefine(std::string name, std::string literal,
```
[the remark this answers is not in the document]
Comment (kornilova-l) [reply bindgen/ir/IR.h]: `LiteralDefine` always have a type now, so I can simply delete the method without a `type`

File: bindgen/defines/DefineFinder.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+#include <clang/Lex/MacroInfo.h>
+#include <llvm/ADT/APInt.h>
+#include <llvm/ADT/APSInt.h>
+
+DefineFinder::DefineFinder(IR &ir, const clang::CompilerInstance &compiler,
+                           clang::Preprocessor &pp)
+    : ir(ir), compiler(compiler), pp(pp) {}
+
+void DefineFinder::MacroDefined(const clang::Token &MacroNameTok,
+                                const clang::MacroDirective *MD) {
```
[the remark this answers is not in the document]
Comment (kornilova-l) [reply bindgen/defines/DefineFinder.cpp]: The names are from the overriden methods, but they can be renamed

**Comment by kornilova-l** [comment]
I couldn't find a function that would expand a define.
My `expandDefine` implementation is able to follow only a simple chain of defines:
```c
#define INT 42
#define INT_2 INT
```
But I think it shouldn't be a big problem to write the implementation that handles more difficult cases.

Once we expand a define we need to process it. If it is a literal then there is no problems to embed it in the Scala code, but  if it's not then we need to find a way to check if it's a variable of data type or of pointer type (or something else). Currently I don't know how to do that.

**Comment by kornilova-l** [comment]
Tests in a container fail because clang API changed since version 4.0
I need to update to 5.0v
We also may want to note somewhere that only versions >= 5.0 are supported

**Comment by kornilova-l** [comment]
To check is a define is an alias for a variable maybe it is a good idea to generate C code that uses the value, feed it into FrontendAction and see if there will be errors

**Comment by kornilova-l** [comment]
I think saving all floating point numbers as `native.CDouble` is good enough for now.

**Comment by kornilova-l** [comment]
Maybe some of C literals are not valid literals in Scala, so it also has to be checked

**Comment by jonas** [comment]
We can test against LLVM 4 if you want. It would however be nice to support more recent versions.

**Comment by jonas** [comment]
I suggest to wait with merging this until 0.1 has been tagged.

**Pull request merged**